delivery_auto_refresh/models/sale_order.py

```diff
@@ -8,115 +9,185 @@
-            .sudo()
-            .get_param("delivery_auto_refresh.auto_add_delivery_line")
-        )
+        return self.company_id.sale_auto_add_delivery_line
+
+    def _get_delivery_discount(self, delivery_lines):
+        return delivery_lines[-1:].discount
+
+    def _update_delivery_line(self, delivery_line, price_unit):
+        """Update the existing delivery line"""
+        values = self._prepare_delivery_line_vals(self.carrier_id, price_unit)
+        new_vals = {}
+        for f, val in values.items():
+            field_def = delivery_line._fields.get(f)
+            if isinstance(field_def, (fields.One2many, fields.Many2many)):
+                # Tax is set with a SET command
+                clear = update = False
+                for cmd in val:
+                    if cmd[0] == fields.Command.SET:
+                        if delivery_line[f].ids != cmd[2]:
+                            update = True
+                    else:
+                        clear = True
+                if clear:
+                    new_vals[f] = [fields.Command.CLEAR] + val
+                elif update:
+                    new_vals[f] = val
+            elif isinstance(field_def, fields.Many2one):
+                if delivery_line[f].id != val:
+                    new_vals[f] = val
+            elif f == "sequence":
+                # sequence is last sequence + 1. As the delivery line already
+                # exists, substract 1
+                if delivery_line[f] != val - 1:
+                    new_vals[f] = val
+            elif delivery_line[f] != val:
+                new_vals[f] = val
+        if new_vals:
+            delivery_line.write(new_vals)
 
     def _auto_refresh_delivery(self):
         self.ensure_one()
-        # Make sure that if you have removed the carrier, the line is gone
-        if self.state in {"draft", "sent"}:
-            # Context added to avoid the recursive calls and save the new
-            # value of carrier_id
-            self.with_context(auto_refresh_delivery=True)._remove_delivery_line()
-        if self._get_param_auto_add_delivery_line() and self.carrier_id:
-            if self.state in {"draft", "sent"}:
-                price_unit = self.carrier_id.rate_shipment(self)["price"]
-                if not self.is_all_service:
-                    self._create_delivery_line(self.carrier_id, price_unit)
-                self.with_context(auto_refresh_delivery=True).write(
-                    {"recompute_delivery_price": False}
-                )
-
-    @api.model
-    def create(self, vals):
-        """Create or refresh delivery line on create."""
-        order = super().create(vals)
-        order._auto_refresh_delivery()
-        return order
+        if self.env.context.get("auto_refresh_delivery"):
+            return
+        if not self._is_auto_add_delivery_line():
+            return
+        if self.state not in ("draft", "sent"):
+            return
```

Comment:
Why not compacting them for less lines and better readability?

```suggestion
        if (
            self.env.context.get("auto_refresh_delivery")
            or not self._is_auto_add_delivery_line()
            or self.state not in ("draft", "sent")
        ):
            return
```